proposals/patterns3.md

```diff
@@ -0,0 +1,253 @@
+
+```antlr
+primary_pattern
+    : parenthesized_pattern
+    ;
+parenthesized_pattern
+    : '(' pattern ')'
+    ;
+```
+
+## Relational Patterns
+
+Relational patterns permit the programmer to express that an input value must satisfy a relational constraint when compared to a constant value:
+
+``` C#
+    public static LifeStage LifeStageAtAge(int age) => age switch
+    {
+        < 0 =>  LiftStage.Prenatal,
+        < 2 =>  LifeStage.Infant,
+        < 4 =>  LifeStage.Toddler,
+        < 6 =>  LifeStage.EarlyChild,
+        < 12 => LifeStage.MiddleChild,
+        < 20 => LifeStage.Adolescent,
+        < 40 => LifeStage.EarlyAdult,
+        < 65 => LifeStage.MiddleAdult,
+        _ =>    LifeStage.LateAdult,
+    };
+```
+
+We imagine supporting `<`, `<=`, `>`, and `>=` patterns on all of the built-in types that support such binary relational operators with two operands of the same type in an expression. Specifically, we support all of these relational patterns for `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, and `decimal`.
+
+```antlr
+primary_pattern
+    : relational_pattern
+    ;
+relational_pattern
+    : '<' relational_expression
+    | '<=' relational_expression
+    | '>' relational_expression
+    | '>=' relational_expression
+    ;
+```
+
+The expression is required to evaluate to a constant value.  It is an error if that constant value is `double.NaN` or `float.NaN`.  It is an error if the expression is a null constant and the relational operator is `<`, `<=`, `>`, or `>=`.
+
+When the input is a type for which a suitable built-in binary relational operator is defined that is applicable with the input as its left operand and the given constant as its right operand, the evaluation of that operator is taken as the meaning of the relational pattern.  Otherwise we convert the input to the type of the expression using an explicit nullable or unboxing conversion.  It is a compile-time error if no such conversion exists.  The pattern is considered not to match if the conversion fails.  If the conversion succeeds then the result of the pattern-matching operation is the result of evaluating the expression `e OP v` where `e` is the converted input, `OP` is the relational operator, and `v` is the constant expression.
+
+## Pattern Combinators
+
+Pattern *combinators* permit matching both of two different patterns using `and` (this can be extended to any number of patterns by the repeated use of `and`), either of two different patterns using `or` (ditto), or the *negation* of a pattern using `not`.
+
+A common use of a combinator will be the idiom
+
+``` c#
+if (e is not null) ...
+```
+
+More readable than the current idiom `e is object`, this pattern clearly expresses that one is checking for a non-null value.
```

Comment:
Worth stating that this is a pure null check?

Reply:
The `null` pattern is already a pure null check.  Placing it under a negation doesn't change that.  In any case, there is no specification of the concept of a pure null check, so there would be no term whose definition could be referenced.